### PR TITLE
IO-702 FileUtils.forceDelete does not delete invalid links. Testcase …

### PR DESCRIPTION
**Pull request by boris-unckel**
…and fix of check.
Testcase and first fix for [https://issues.apache.org/jira/browse/IO-702](https://issues.apache.org/jira/browse/IO-702)

### CODE REVIEW

**Comment by boris-unckel** [comment]
@garydgregory
The deleteDirectory story is still not final, but with different root cause. Please review the testcase and first fix. Do you have an idea whats wrong with forceDelete?

File: src/main/java/org/apache/commons/io/FileUtils.java [hunk 1]
```diff
@@ -309,6 +309,9 @@ public static void cleanDirectory(final File directory) throws IOException {
         if (!causeList.isEmpty()) {
             throw new IOExceptionList(causeList);
         }
+        if (listFiles(directory, null).length > 0) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/FileUtils.java:312]: Hi @boris-unckel 

Thank you for your PR.

That is an expensive way to test for emptiness if there are a lot of files. Please rebase on master and call `isEmptyDirectory(File)` in this class.

TY.

File: src/test/java/org/apache/commons/io/FileUtilsDeleteDirectoryBaseTestCase.java [hunk 2]
```diff
@@ -194,6 +194,29 @@ public void testDeleteParentSymlink2() throws Exception {
         assertEquals(1, randomDirectory.list().length, "Contents of sym link should not have been removed");
     }
 
+    @Test
+    public void testDeleteInvalidLinks() throws Exception {
+        File AFile = new File(top, "realParentDirA");
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/FileUtilsDeleteDirectoryBaseTestCase.java:199]: These local variable names are unconventional, `AFile` -> `fileA` as local variable names should start with a lower case chaarcter.

File: src/main/java/org/apache/commons/io/FileUtils.java [hunk 3]
```diff
@@ -1342,6 +1345,9 @@ public static void forceDelete(final File file) throws IOException {
             // didn't find a file to delete.
             throw new FileNotFoundException("File does not exist: " + file);
         }
+        if (file.exists()) {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/FileUtils.java:1348]: This belongs in the unit test.

File: src/main/java/org/apache/commons/io/FileUtils.java [hunk 4]
```diff
@@ -309,6 +309,9 @@ public static void cleanDirectory(final File directory) throws IOException {
         if (!causeList.isEmpty()) {
             throw new IOExceptionList(causeList);
         }
+        if (listFiles(directory, null).length > 0) {
+            throw new IOException("Directory " + directory + " is not empty, cleaning attempt failed.");
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/FileUtils.java:313]: This belongs in the unit test IMO.

File: src/test/java/org/apache/commons/io/FileUtilsDeleteDirectoryBaseTestCase.java (hunk 2 above)
Comment (boris-unckel) [reply src/test/java/org/apache/commons/io/FileUtilsDeleteDirectoryBaseTestCase.java]: thnx, corrected.

File: src/main/java/org/apache/commons/io/FileUtils.java (hunk 4 above)
Comment (boris-unckel) [reply src/main/java/org/apache/commons/io/FileUtils.java]: removed

File: src/main/java/org/apache/commons/io/FileUtils.java (hunk 3 above)
Comment (boris-unckel) [reply src/main/java/org/apache/commons/io/FileUtils.java]: removed

File: src/main/java/org/apache/commons/io/FileUtils.java (hunk 1 above)
Comment (boris-unckel) [reply src/main/java/org/apache/commons/io/FileUtils.java]: obsolete - it's removed, see later comment

**Comment by boris-unckel** [comment]
@garydgregory 
Now it fails again in deleteDirectory (which is not the cause)
`Error:  org.apache.commons.io.FileUtilsDeleteDirectoryLinuxTestCase.testDeleteInvalidLinks  Time elapsed: 0.003 s  <<< ERROR!
java.nio.file.DirectoryNotEmptyException: /tmp/junit16834268871756684963/realParentDirC
	at java.base/sun.nio.fs.UnixFileSystemProvider.implDelete(UnixFileSystemProvider.java:247)
	at java.base/sun.nio.fs.AbstractFileSystemProvider.delete(AbstractFileSystemProvider.java:105)
	at java.base/java.nio.file.Files.delete(Files.java:1141)
	at org.apache.commons.io.FileUtils.delete(FileUtils.java:1170)
	at org.apache.commons.io.FileUtils.deleteDirectory(FileUtils.java:1189)
	at org.apache.commons.io.FileUtilsDeleteDirectoryBaseTestCase.testDeleteInvalidLinks(FileUtilsDeleteDirectoryBaseTestCase.java:216)`

IMO This stacktrace is misleading because originally forceDelete fails to delete the broken link. Maybe it's the Visitor in the nio TreeWalker.

**Comment by boris-unckel** [comment]
@garydgregory
Fixed the DeletingPathVisitor.visitFile with an additional check for symlinks. Please review again.

**Pull request merged**